map/mwm_url.cpp

```diff
@@ -86,53 +116,58 @@ ParsedMapApi::ParsingResult ParsedMapApi::SetUriAndParse(string const & url)
-  string const & path = uri.GetPath();
-  bool const isRoutePath = path == "route";
-  if ((scheme != "mapswithme" && scheme != "mwm" && scheme != "mapsme") ||
-      (path != "map" && !isRoutePath))
+  switch (URLTypeFrom(uri))
   {
-    return ParsingResult::Incorrect;
-  }
-
-  if (isRoutePath)
-  {
-    m_routePoints.clear();
-    vector<string> pattern{kSourceLatLon, kSourceName, kDestLatLon, kDestName, kRouteType};
-    if (!uri.ForEachKeyValue(bind(&ParsedMapApi::RouteKeyValue, this, _1, _2, ref(pattern))))
+    case ApiURLType::Incorrect:
       return ParsingResult::Incorrect;
-
-    if (pattern.size() != 0)
-      return ParsingResult::Incorrect;
-
-    if (m_routePoints.size() != 2)
+    case ApiURLType::Map:
     {
-      ASSERT(false, ());
-      return ParsingResult::Incorrect;
+      vector<ApiPoint> points;
+      if (!uri.ForEachKeyValue(bind(&ParsedMapApi::AddKeyValue, this, _1, _2, ref(points))))
+        return ParsingResult::Incorrect;
+
+      if (points.empty())
+        return ParsingResult::Incorrect;
+
+      ASSERT(m_bmManager != nullptr, ());
+      UserMarkControllerGuard guard(*m_bmManager, UserMarkType::API_MARK);
+      for (auto const & p : points)
+      {
+        m2::PointD glPoint(MercatorBounds::FromLatLon(p.m_lat, p.m_lon));
+        ApiMarkPoint * mark = static_cast<ApiMarkPoint *>(guard.m_controller.CreateUserMark(glPoint));
+        mark->SetName(p.m_name);
+        mark->SetID(p.m_id);
+        mark->SetStyle(style::GetSupportedStyle(p.m_style, p.m_name, ""));
+      }
+
+      return ParsingResult::Map;
     }
+    case ApiURLType::Route:
+    {
+      m_routePoints.clear();
+      vector<string> pattern{kSourceLatLon, kSourceName, kDestLatLon, kDestName, kRouteType};
+      if (!uri.ForEachKeyValue(bind(&ParsedMapApi::RouteKeyValue, this, _1, _2, ref(pattern))))
+        return ParsingResult::Incorrect;
 
-    return ParsingResult::Route;
-  }
-
-  vector<ApiPoint> points;
-  if (!uri.ForEachKeyValue(bind(&ParsedMapApi::AddKeyValue, this, _1, _2, ref(points))))
-    return ParsingResult::Incorrect;
+      if (pattern.size() != 0)
+        return ParsingResult::Incorrect;
 
-  if (points.empty())
-    return ParsingResult::Incorrect;
+      if (m_routePoints.size() != 2)
+      {
+        ASSERT(false, ());
+        return ParsingResult::Incorrect;
+      }
 
-  ASSERT(m_bmManager != nullptr, ());
-  UserMarkControllerGuard guard(*m_bmManager, UserMarkType::API_MARK);
-  for (auto const & p : points)
-  {
-    m2::PointD glPoint(MercatorBounds::FromLatLon(p.m_lat, p.m_lon));
-    ApiMarkPoint * mark = static_cast<ApiMarkPoint *>(guard.m_controller.CreateUserMark(glPoint));
-    mark->SetName(p.m_name);
-    mark->SetID(p.m_id);
-    mark->SetStyle(style::GetSupportedStyle(p.m_style, p.m_name, ""));
+      return ParsingResult::Route;
+    }
+    case ApiURLType::Search:
+      SearchRequest request;
+      if (!uri.ForEachKeyValue(bind(&ParsedMapApi::SearchKeyValue, this, _1, _2, ref(request))))
```

Comment:
Ah, see now, SearchKeyValue is non-trivial. Okay, feel free to leave this code as is.